pyxem/hyperspy_extension.yaml

```diff
@@ -18,109 +18,109 @@
 
 signals:
   Diffraction1D:
-    signal_type: "Diffraction1D"
+    signal_type: diffraction1d
     signal_dimension: 1
     dtype: real
     lazy: False
     module: pyxem.signals.diffraction1d
   Diffraction2D:
-    signal_type: "Diffraction2D"
+    signal_type: diffraction2d
     signal_dimension: 2
     dtype: real
     lazy: False
     module: pyxem.signals.diffraction2d
   LazyDiffraction1D:
-    signal_type: "LazyDiffraction1D"
+    signal_type: diffraction1d
     signal_dimension: 1
     dtype: real
     lazy: True
     module: pyxem.signals.diffraction1d
   LazyDiffraction2D:
-    signal_type: "LazyDiffraction2D"
+    signal_type: diffraction2d
     signal_dimension: 2
     dtype: real
     lazy: True
     module: pyxem.signals.diffraction2d
   ElectronDiffraction1D:
-    signal_type: "ElectronDiffraction1D"
+    signal_type: electron_diffraction1d
     signal_dimension: 1
     dtype: real
     lazy: False
     module: pyxem.signals.electron_diffraction1d
   ElectronDiffraction2D:
-    signal_type: "ElectronDiffraction2D"
+    signal_type: electron_diffraction2d
     signal_dimension: 2
     dtype: real
     lazy: False
     module: pyxem.signals.electron_diffraction2d
   LazyElectronDiffraction1D:
-    signal_type: "LazyElectronDiffraction1D"
+    signal_type: electron_diffraction1d
     signal_dimension: 1
     dtype: real
     lazy: True
     module: pyxem.signals.electron_diffraction1d
   LazyElectronDiffraction2D:
-    signal_type: "LazyElectronDiffraction2D"
+    signal_type: electron_diffraction2d
     signal_dimension: 2
     dtype: real
     lazy: True
     module: pyxem.signals.electron_diffraction2d
   DiffractionVectors:
-    signal_type: "DiffractionVectors"
-    signal_dimension: None
+    signal_type: diffraction_vectors
+    signal_dimension: 0
     dtype: real
     lazy: False
     module: pyxem.signals.diffraction_vectors
   TemplateMatchingResults:
-    signal_type: "TemplateMatchingResults"
-    signal_dimension: None
+    signal_type: template_matching
+    signal_dimension: 2
     dtype: real
     lazy: False
     module: pyxem.signals.indexation_results
   VectorMatchingResults:
-    signal_type: "VectorMatchingResults"
-    signal_dimension: None
+    signal_type: vector_matching
+    signal_dimension: 0
     dtype: real
     lazy: False
     module: pyxem.signals.indexation_results
   CrystallographicMap:
-    signal_type: "CrystallographicMap"
-    signal_dimension: None
+    signal_type: crystallographic_map
+    signal_dimension: 1
```

Comment:
Not sure if I understand correctly what these signals are designed for and if their signal dimension are correct here. Same for `DiffractionVectors`.

Reply:
Both CrystallographicMap and DiffractionVectors don't really fit the hyperspy signal type very well.

The dimension of CrystallographicMap is the navigation dimension of the original ElectronDiffraction2D object it came from. Probably we should have a 1D and 2D variation of it, but the medium term plan is to get rid of this class all together and make it functionality of pyxem/orix instead, which does not depend on hyperspy. (See https://github.com/pyxem/orix/issues/48)

DiffractionVectors has the problem that it is typically made by doing peak finding on an ElectronDiffraction2D object so it's signal dimension is either 2 if there are the same number of peaks in every diffraction pattern or, more typically, it is ragged.

I'm not really sure how we should solve that in a more general way.

Reply:
The actual CrystalMap PR is here : https://github.com/pyxem/orix/pull/49

@hakonanes and @pc494 what do we think the timeline is like on the new CrystalMap class for orix? It would be good to get rid of this thing from pyxem if we can, but no real rush.

Reply:
I hope to push basic `CrystalMap` class with attributes and indexing via `__getitem__()` (which, in my opinion, is where the power in such a class lies, i.e. masking e.g. an EBSD scan by grain ID, or any other imaginable property) within the next couple days. As far as I can tell, I only have the `__getitem__()` method left. Then we need a back-and-forth on best practice, after which I have to write tests. Difficult to estimate a finishing date, but that is where I'm at.

Reply:
Actually, `CrystallographicMap` dimension is explicitly set to 1 and its data is of type `'o'`, the signal dimension seems to be used mainly for slicing using the `isig` syntax, so I guess this would work fine as it is.

The `DiffractionVectors` case should now work both case, if this is a ragged or 2D signal.

Reply:
Thanks @ericpre - we had avoided doing what you've done with DiffractionVectors & DiffractionVectors2D there because we're planning on re-factoring all of the DiffractionVectors code to separate the 2D detector from 3D Fourier space, which is important for accounting for the curvature of the Ewald sphere properly and also for assessing the diffraction data in 3D, which we hope to support by the summer.

In https://github.com/pyxem/pyxem/pull/485 you can see the 3 categories of vector I was anticipating supporting: DetectorCoordinates2D, DiffractionVectors2D, and DiffractionVectors3D.

Here my dimensionality indicators are referring to the number of coordinates per vector in each of those 3 cases. In terms of hyperspy signal dimensions when generally applied to mapping or tilt-series or time-series data the number of vectors per navigation index will typically change, so these would typically be ragged arrays, but then there is the "special case" where the sample is special or there is only 1 diffraction pattern when the map method would signal dimensions equal to the number of coordinates describing the vector.

If we do what you've done here, then we'd end up with 6 different Vector classes, which seemed like a not great solution.

But I appreciate that what we've done is not at all consistent with how a hyperspy Signal "should" be and that what you're suggesting is.

Really I think we need to come up with a more general solution that encompasses signals of this type we've ended up with here i.e. where there is a natural dimensionality to the data (the vector dimensionality) but the Signal will often be ragged. This could also be used for the output of 1D peak finding for example. Do you have any thoughts on a hyperspy-esque way to do that?

This might interest @francisco-dlp I guess.

Reply:
Reflecting on this some more, I suppose it might be better to only have DiffractionVectors class with the 3D coordinates as the primary signal, since that is the actual reciprocal space and store the coordinates associated with the detector as attributes of  that object.

Perhaps DiffractionVectors just shouldn't be a sub-class of BaseSignal? But I like being able to use the hyperspy map() on ragged signals...

Reply:
Would it be ok if in this PR, we stick to fixing the signal registration and avoid sorting out the issue with `DiffractionVectors`?

Even if the diffraction vectors issue needs to be sorted out and more generally it is an interesting problem to solve, my motivation for this PR is to be able to use pyxem signal with normal hyperspy functions more easily in the short term.

At the moment, I don't have a clear idea of this diffraction vectors issue involved but below is a list of consideration which I think should be relevant:
- ragged array are BaseSignal with signal dimension of 0 (to my understanding)
- using the signal registration mechanism allow to assign to the correct signal dimensionality "on the fly", as it is done in this PR (to return either `DiffractionVectors` or `DiffractionVectors2D`)
- using subclasses of `DiffractionVectors` with specific signal dimension would allow the use of more efficient functions or appropriate plotting

> Perhaps DiffractionVectors just shouldn't be a sub-class of BaseSignal? But I like being able to use the hyperspy map() on ragged signals...

I was also wondering if it should be a subclass of `BaseSignal` (particularly for `CrystallographicMap`) but as long as there is a need to have a navigation dimension, I think that there are actually only benefits to keep it like this and I can't think of any drawbacks.

In this PR, I have added `DiffractionVectors2D` so that pyxem works in the same way as it is currently - as far as I understand - so that saving/loading are still working fine. I guess the API is not fully stabilised, so in the worst case scenario, the `DiffractionVectors2D` could be dropped at a latter staged without worrying too much about it? Moreover, if the main source of creation of `DiffractionVectors2D` signal is as an output of `find_peaks`, it would be rarely used.

Reply:
Yes I agree with these being things for separate PRs, just made me think of it and I thought you might already know how to do it better than me so wanted to bring it up.

Similarly to DiffractionVectors changing again in future (which I'm fine with), CrystallographicMap will be removed in future and imported from elsewhere in due course so I'm also quite relaxed about that for now.